src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs

```diff
@@ -30,179 +32,165 @@ protected MessageBody(Frame context)
-            Stream destination,
-            CancellationToken cancellationToken = default(CancellationToken))
+        public virtual async Task ConsumeAsync(CancellationToken cancellationToken = default(CancellationToken))
         {
+            Exception error = null;
+
             try
             {
-                await destinationTask;
+                while (true)
+                {
+                    var result = await _context.RequestBodyPipe.Reader.ReadAsync();
+                    _context.RequestBodyPipe.Reader.Advance(result.Buffer.End);
+
+                    if (result.IsCompleted)
+                    {
+                        return;
+                    }
+                }
             }
-            finally
+            catch (Exception ex)
             {
-                ConsumedBytes(bytesConsumed);
+                error = ex;
+                throw;
             }
-
-            var peekTask = PeekAsync(cancellationToken);
-
-            if (!peekTask.IsCompleted)
+            finally
             {
-                TryProduceContinue();
+                _context.RequestBodyPipe.Reader.Complete(error);
             }
-
-            await CopyToAsyncPeekAwaited(peekTask, destination, cancellationToken);
         }
 
-        public Task Consume(CancellationToken cancellationToken = default(CancellationToken))
+        protected void OnData(ReadableBuffer readableBuffer)
         {
-            while (true)
-            {
-                var task = PeekAsync(cancellationToken);
-                if (!task.IsCompleted)
-                {
-                    TryProduceContinue();
+            var writableBuffer = _context.RequestBodyPipe.Writer.Alloc(0);
 
-                    // Incomplete Task await result
-                    return ConsumeAwaited(task, cancellationToken);
-                }
-                else
-                {
-                    // ValueTask uses .GetAwaiter().GetResult() if necessary
-                    if (task.Result.Count == 0)
-                    {
-                        // Completed Task, end of stream
-                        return TaskCache.CompletedTask;
-                    }
-
-                    ConsumedBytes(task.Result.Count);
-                }
+            try
+            {
+                writableBuffer.Append(readableBuffer);
             }
-        }
-
-        private async Task ConsumeAwaited(ValueTask<ArraySegment<byte>> currentTask, CancellationToken cancellationToken)
-        {
-            while (true)
+            finally
             {
-                var count = (await currentTask).Count;
-
-                if (count == 0)
-                {
-                    // Completed Task, end of stream
-                    return;
-                }
-
-                ConsumedBytes(count);
-                currentTask = PeekAsync(cancellationToken);
+                writableBuffer.Commit();
+                _currentWritableBuffer = writableBuffer;
```

Comment:
I could use the same pattern from `LibuvConnection`: https://github.com/aspnet/KestrelHttpServer/blob/41f1922502f475802c6cf3ba2cbe3653ab5d00fd/src/Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv/Internal/LibuvConnection.cs#L140-L147

And call `FlushAsync()` in `OnData()`. I'll have to care about backpressure anyways, when I add the timeout.